lib/rlpbwt_bv.cpp

```diff
@@ -0,0 +1,1292 @@
+                // lcp array in order to calculate how many rows are matching at
+                // this point of the query, updating the end index
+                while (end_tmp < this->height &&
+                       (i + 1) - this->cols[i + 1].lcp[end_tmp] <= curr_beg) {
+                    end_tmp++;
+                }
+
+                // update in second case is complete
+                curr_index = curr_tmp;
+                end_index = end_tmp;
+                if (verbose) {
+                    std::cout << "end case 3 curr index " << curr_index
+                              << ", end index " << end_index << "\n";
+                }
+            }
+        }
+        if (verbose) {
+            std::cout << "after at " << i << " from " << curr_index << " to "
+                      << end_index
+                      << "\n";
+        }
+    }
+    // check/save matches at the end (iff length > 1)
+    if (curr_index < end_index) {
+        curr_len = end_index - curr_index;
+        if (verbose) {
+            std::cout << "match at (" << curr_beg << ", " << query.size() - 1
+                      << ") with " << curr_len << " haplotypes \n";
+        }
+        if (query.size() - curr_beg > 1) {
+            matches.basic_matches.emplace_back(curr_len,
+                                               query.size() - curr_beg,
+                                               query.size() - 1);
+        }
+    }
+    return matches;
+}
+
+unsigned int
+rlpbwt_bv::lf(unsigned int col_index, unsigned int index, char symbol,
+              bool verbose) const {
+    // obtain "u" and "v"
+    auto uv = uvtrick(col_index, index);
+    if (verbose) {
+        std::cout << "at col: " << col_index << " and index " << index
+                  << " with symbol " << symbol << "\n";
+        std::cout << uv.first << " " << uv.second << " with c: "
+                  << this->cols[col_index].count_0 << "\n";
+    }
+    // computer lf-mapping as Durbin's w(i, s)
+    if (symbol == '0') {
+        return uv.first;
+    } else {
+        return this->cols[col_index].count_0 + uv.second;
+    }
+}
+
+std::pair<unsigned int, unsigned int>
+rlpbwt_bv::uvtrick(unsigned int col_index, unsigned int index) const {
+    // if index is 0 u = v = 0
+    if (index == 0) {
+        return {0, 0};
+    }
+
+    // number of previous zeros and ones
+    unsigned int prev_u = 0;
+    unsigned int prev_v = 0;
+    // runs of previous zeros and ones in respective bitvectors
+    unsigned int run_prev_u = 0;
+    unsigned int run_prev_v = 0;
+    // offset to finally consider inside the run
+    unsigned int offset = 0;
+
+    // compute the run of the index using rank over runs bitvector
+    unsigned run = this->cols[col_index].rank_runs(index);
+    // two main cases:
+    // - run > 1
+    // - run = 0 or run = 1 (we separate this two cases due to some errors using
+    //   select)
+    if (run > 1) {
```

Comment:
This is a very long if with nested ifs. I would rewrite it as:
```
if(run%2 == 0){
   if(run == 0) return ...;
}else{
   if(run ==1) return ...;
}
```

You could also do:
```
if(run==0)
if(run==1)
if(run%2==0){
}else{
}
```
which has one more test to do every time, but I don't know if this might help branch prediction or not.